File: cbt_serialize.cpp
#include "cbt_serialize.h"
#include "serialize.h"
#include <iostream>
#include <queue>
#include <string>

using namespace std;

// Вспомогательная функция для сериализации дерева в текстовом формате
void CbtSerializer::serializeTree(ostream& os, CompleteBinaryTree::TreeNode* node) {
    if (!node) {
        // Для текстового формата используем обычный вывод
        os << "null\n";
        return;
    }
    os << node->data << "\n";
    serializeTree(os, node->left);
    serializeTree(os, node->right);
}

// Вспомогательная функция для десериализации дерева из текстового формата
CompleteBinaryTree::TreeNode* CbtSerializer::deserializeTree(istream& is) {
    string token;
    if (!getline(is, token) || token.empty()) return nullptr;
    
    if (token == "null") return nullptr;
    
    try {
        int value = stoi(token);
        CompleteBinaryTree::TreeNode* node = new CompleteBinaryTree::TreeNode(value);
        node->left = deserializeTree(is);
        node->right = deserializeTree(is);
        return node;
    } catch (...) {
        return nullptr;
    }
}

// Вспомогательная функция для сериализации дерева в бинарном формате
void CbtSerializer::serializeTreeBinary(ostream& os, CompleteBinaryTree::TreeNode* node) {
    if (!node) {
        int marker = -1;
        os.write(reinterpret_cast<const char*>(&marker), sizeof(marker));
        return;
    }
    os.write(reinterpret_cast<const char*>(&node->data), sizeof(node->data));
    serializeTreeBinary(os, node->left);
    serializeTreeBinary(os, node->right);
}

// Вспомогательная функция для десериализации дерева из бинарного формата
CompleteBinaryTree::TreeNode* CbtSerializer::deserializeTreeBinary(istream& is) {
    int value;
    is.read(reinterpret_cast<char*>(&value), sizeof(value));
    if (is.fail() || value == -1) return nullptr;
    
    CompleteBinaryTree::TreeNode* node = new CompleteBinaryTree::TreeNode(value);
    node->left = deserializeTreeBinary(is);
    node->right = deserializeTreeBinary(is);
    return node;
}

void CbtSerializer::saveToFile(const CompleteBinaryTree& tree, const string& filename) {
    ofstream file(filename);
    if (!file) return;
    serializeTree(file, tree.getRoot_test());
}

void CbtSerializer::loadFromFile(CompleteBinaryTree& tree, const string& filename) {
    ifstream file(filename);
    if (!file) return;
    
    // Очищаем существующее дерево
    tree.clear();
    
    // Десериализуем новое дерево
    CompleteBinaryTree::TreeNode* newRoot = deserializeTree(file);
    
    // Устанавливаем новый корень
    tree.setRoot(newRoot);
}

void CbtSerializer::saveToBinaryFile(const CompleteBinaryTree& tree, const string& filename) {
    ofstream file(filename, ios::binary);
    if (!file) return;
    serializeTreeBinary(file, tree.getRoot_test());
}

void CbtSerializer::loadFromBinaryFile(CompleteBinaryTree& tree, const string& filename) {
    ifstream file(filename, ios::binary);
    if (!file) return;
    
    // Очищаем существующее дерево
    tree.clear();
    
    // Десериализуем новое дерево
    CompleteBinaryTree::TreeNode* newRoot = deserializeTreeBinary(file);
    
    // Устанавливаем новый корень
    tree.setRoot(newRoot);
}

File: ds/cbt_serialize.go
package ds

import (
	"os"
)

type CbtSerializer struct{}

func NewCbtSerializer() *CbtSerializer {
	return &CbtSerializer{}
}

func (s *CbtSerializer) SaveToFile(tree *CompleteBinaryTree, filename string) error {
	return SaveCBTText(filename, tree.Root)
}

func (s *CbtSerializer) LoadFromFile(tree *CompleteBinaryTree, filename string) error {
	root, err := LoadCBTText(filename)
	if err != nil {
		return err
	}
	tree.Root = root
	tree.Size = tree.countNodes(tree.Root)
	return nil
}

func writeBinaryRec(n *CBTNode, f *os.File) error {
	if n == nil {
		return WriteInt32(f, -1)
	}
	if err := WriteInt32(f, n.Data); err != nil {
		return err
	}
	if err := writeBinaryRec(n.Left, f); err != nil {
		return err
	}
	return writeBinaryRec(n.Right, f)
}

func (s *CbtSerializer) SaveToBinaryFile(tree *CompleteBinaryTree, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	return writeBinaryRec(tree.Root, file)
}

func readBinaryRec(f *os.File) (*CBTNode, error) {
	val, err := ReadInt32(f)
	if err != nil {
		return nil, err
	}
	if val == -1 {
		return nil, nil
	}
	node := &CBTNode{Data: val}
	node.Left, err = readBinaryRec(f)
	if err != nil {
		return nil, err
	}
	node.Right, err = readBinaryRec(f)
	if err != nil {
		return nil, err
	}
	return node, nil
}

func (s *CbtSerializer) LoadFromBinaryFile(tree *CompleteBinaryTree, filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	tree.Clear()
	root, err := readBinaryRec(file)
	if err != nil {
		return err
	}
	tree.Root = root
	tree.Size = tree.countNodes(tree.Root)
	return nil
}